Tgui stat panel

PowerfulBacon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

**Fix chat HARDER ports:**
 - https://github.com/tgstation/tgstation/pull/54030/files
 - https://github.com/tgstation/tgstation/pull/54517

## Important
client.verbs +=, client.verbs -= and client.verbs |= is now not allowed and will not function properly. Once this is merged, all PRs will be required to do client.add_verb()

## About The Pull Request

Adds in the beautiful TGUI stat panel.
Actually includes all of the verbs.

Completely depreciates statclick and /datum/verbs

## Why It's Good For The Game

It looks good.
MC no longer lags the server.

**NOTE: TABS WERE NOT SORTED ALPHABETICALLY IN THESE SCREENSHOTS (Added that afterwards)** (actually I haven't fixed that yet even 1 day later)  <-- I have now

Status:
![image](https://user-images.githubusercontent.com/26465327/96755385-8fb11200-13ca-11eb-8428-4a298dc87bb1.png)
![image](https://user-images.githubusercontent.com/26465327/96755500-b53e1b80-13ca-11eb-9847-6b9ff6d03534.png)
![image](https://user-images.githubusercontent.com/26465327/96755542-c129dd80-13ca-11eb-8987-d283e3edebaf.png)
![image](https://user-images.githubusercontent.com/26465327/96755586-cf77f980-13ca-11eb-9678-d99f80f582f8.png)
(Not lagging the game or the server, and updating in real time)
![image](https://user-images.githubusercontent.com/26465327/96755615-db63bb80-13ca-11eb-8a49-4f2b069a16de.png)
![image](https://user-images.githubusercontent.com/26465327/96755628-def74280-13ca-11eb-9183-ab2293c5a127.png)
![image](https://user-images.githubusercontent.com/26465327/96755645-e4548d00-13ca-11eb-88f2-1b7ba77f8a50.png)
![image](https://user-images.githubusercontent.com/26465327/96759734-5a0f2780-13d0-11eb-9fd2-6bce00c95695.png)

## Todo:
 - [x] Make it so you can resize the stat panel.
 - [x] Clean up some ungodly stat code.
 - [x] Fix the single ticket manager issue

## Important things to note / Potential Issues:
This completely removes the stat panel, so if TG chat fails to load the only way to use the fix-chat verb (which barely ever fixes chat) is to type in the command line. I'll port something that adds a message to chat from TG to help with this, but for now we have a message instead.

## Changelog
:cl:
add: TGUI stat panel
del: Old stat panel
del: Depreciates /datum/verbs
del: Depreciates statclick
refactor: Refactored how verbs are added to things
refactor: Refactors stat code
fix: Ports fix fix chat from TG station
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PowerfulBacon committed:
before python script

PowerfulBacon committed:
Ports TGchat, ports TGUI4, ports all the other shit in between

PowerfulBacon committed:
Compiles TGUI

PowerfulBacon committed:
e

PowerfulBacon committed:
Updates the external_rsc_urls

PowerfulBacon committed:
Updates the external RSC URLs to be correct

PowerfulBacon committed:
Updates ui_assets on all UIs

PowerfulBacon committed:
oh shit

PowerfulBacon committed:
Removes accidental TGUI files, makes all interfaces use the new width and height system.

PowerfulBacon committed:
fixes and compiles TGUI errors

PowerfulBacon committed:
Fixes a few UI

PowerfulBacon committed:
Updates TGUI bundle

PowerfulBacon committed:
e

PowerfulBacon committed:
Updates external_src_urls

PowerfulBacon committed:
Uncomments stuff that I think should be uncommented. Fixes MD5 issue

PowerfulBacon committed:
Updates some things

PowerfulBacon committed:
removes duplicate include

PowerfulBacon committed:
push

Archanial committed:
Unfucks cloning (#2579)

* mewhen

* take2

ss13-beebot committed:
Automatic changelog generation for PR #2579 [ci skip]

PowerfulBacon committed:
e

PowerfulBacon committed:
e

PowerfulBacon committed:
e

PowerfulBacon committed:
e

PowerfulBacon committed:
Dixes the stylesheet

PowerfulBacon committed:
Dixes the stylesheet

PowerfulBacon committed:
Update!

PowerfulBacon committed:
Update!

PowerfulBacon committed:
update

PowerfulBacon committed:
sorry

PowerfulBacon committed:
Update beestation.dme

PowerfulBacon committed:
e

PowerfulBacon committed:
d

PowerfulBacon committed:
papershitcode

Commit:
Merge branch 'master' into tgui-4-port

Commit:
Update paper.dm

Commit:
Fixes staging bugs

Commit:
Update arcade.dm

Commit:
Adds names to UIs

Commit:
Fixes GPS, Fixes wires

Commit:
e

Commit:
Merge branch 'master' into tgui-4-port

Commit:
Fixes computer UI and view variables not having proper assets sent

Commit:
Rebuilds TGUI, fixes an issue related to having light window with dark chat if TGUI hasn't been loaded before

Commit:
Fixes crafting

Commit:
Emote colours, seperated settings bar

Commit:
Update tgui_panel.dm

Commit:
Update settings.json

Commit:
Merge branch 'master' into tgui-4-port

Commit:
final compile

Commit:
wtf

Commit:
Updates wires

Commit:
epic

Commit:
Day 1

Commit:
Day 1

Commit:
Update mob.dm

Commit:
Update mob.dm

Commit:
e

Commit:
e

Commit:
Dividers, MC

Commit:
Dividers, MC

Commit:
Verbs

Commit:
Verbs

Commit:
Stat messages

Commit:
Stat messages

Commit:
Finishes the stats for other things

Commit:
Finishes the stats for other things

Commit:
Compiles TGUI

Commit:
Compiles TGUI

Commit:
Nukes /datum/verbs

Commit:
Nukes /datum/verbs

Moccha-Bee commented:
papa bless

Commit:
based

Commit:
based

Commit:
Before pull

Commit:
Before pull

Commit:
Merge branch 'master' into TGUI-stat-panel

Commit:
Removes fucked up merge conflict resolution

Commit:
Undoes the game

Commit:
Update tgui.bundle.js

Commit:
Merge branch 'TGUI-stat-panel' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork into TGUI-stat-panel

Commit:
Update beestation.dme

Commit:
Stat panel now retains its size

Commit:
Fixes tickets not opening

Commit:
Replaces lists with GENERATE_STAT_TEXT

Commit:
Cleans the code, adds spells to stat panel

Commit:
Ports fixes fix chat

Commit:
Undefines low memory mode

PowerfulBacon commented:
Ready

Commit:
Fixes runtimes

Archanial commented:
okay but how does it look with proposed 17x15?

PowerfulBacon commented:
> 
> 
> okay but how does it look with proposed 17x15?

Should look pretty similar, the tab selection might scroll though but that can be changed to wrap to new line if needed. Also I think I forgot to update the alphabetical order bars

ike709 commented:
![image](https://user-images.githubusercontent.com/5714543/97094766-9be0dd80-161d-11eb-8470-2bdb6f8c59a1.png)

Status panel's values have no spacing

ike709 commented:
IC notes seem to not work at all.

ike709 commented:
CPU usage is maxing out and I think it's this PR's fault.

ike709 commented:
A bunch of verbs are supposedly missing.

ike709 commented:
Several players report layout issues but I have no screencaps.

ike709 commented:
Confirmed this was the cause of CPU issues.

PowerfulBacon commented:
TODO:
 - Put sending messages into middleware.
 - Stat panel subsystem

Commit:
Less server killing, tab sorting

Commit:
Updates TGUI build

Commit:
Fixes missing verbs, updates the TGUI bundle.

Commit:
Cool juice

PowerfulBacon commented:
Recommended with https://github.com/BeeStation/BeeStation-Hornet/pull/2668

PowerfulBacon commented:
Player panel shows under admin for non admins but cannot be used.
Chat seems to be cut off for almost all players (Chat is slightly too big for non native resolution users).
There is still no gab between the text stats
https://cdn.discordapp.com/attachments/427337870844362755/770600184614617098/Screenshot_1.png
Gas scan is missing
Debug admin log on finding ticket with ID [id]

ike709 commented:
Conflicts.

Commit:
Merge branch 'master' into TGUI-stat-panel

Commit:
e

Commit:
Works now

Commit:
Merge branch 'master' into TGUI-stat-panel

Commit:
e

Cenrus commented:
Observe, Toggle HUD, check health, and gas scan verbs are missing from ghosts

PowerfulBacon commented:
On it

Commit:
Removes debug message

Wilchenx commented:
![image](https://user-images.githubusercontent.com/25509881/98472118-223a1980-21f1-11eb-9472-31b7e85d8343.png)
Normal players can see player panel which should not be visible, button itself doesn't do anything without perms.

PowerfulBacon commented:
Huh that's an odd one

Commit:
Update Panel.js

PowerfulBacon commented:
Seemed to cause a lot of lag too, might have been unrelated but highly unlikely.

Commit:
Update all_nodes.dm

ike709 commented:
How's this going? Also, conflicts.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into TGUI-stat-panel

PowerfulBacon commented:
Going through some fixes now, fixing the data lag *might* be a little more hard

PowerfulBacon commented:
Fixed all cases of missing verbs

PowerfulBacon commented:
Fixed player panel bug

PowerfulBacon commented:
Last thing to do is move the transfer of verb data to the verb on gain rather than on switch tab to remove data lag.

ike709 commented:
Conflicts

Commit:
a

PowerfulBacon commented:
Data lag should be slightly less, large tabs will now update slower and the stat subsystem is background

Commit:
MMMM juicy changes

Commit:
Fixes the alert popup at roundstart

Commit:
Antagonist Popup Fixes

Commit:
Another fix

Commit:
Forces updates, fixes multiple of the same tab

Commit:
Cool

Commit:
Update client_procs.dm

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into TGUI-stat-panel

Commit:
Fixes deadmin being slow, bluescreening chat and something else

Commit:
Fixes readmin deadmin not showing verbs instantly

PowerfulBacon commented:
useLocalState(context, 'number', settings.statSize); can be null?

PowerfulBacon commented:
Object verbs were lost

IllotAFV commented:
OOC is missing a Ghost tab, OOC also doesnt have Observe tab.

Commit:
based?

ike709 commented:
Conflicts

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Update job.dm

Commit:
Revert "Update job.dm"

This reverts commit 97ee7d2bae09798ab58d5cfc5afba2600ea06ae2.

Commit:
easef

Commit:
Revert "easef"

This reverts commit 6bcedf9c3e39165214a0b0a7b3831f461bbf61a9.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

PowerfulBacon commented:
ok this is a bit boring but going to try and get it done today. Should be easy to grab the verbs of things the player is wearing since I just realized its mob.contents

Commit:
Adds in object verbs

Commit:
Updates stuff

Commit:
Update tgui_panel.dm

PowerfulBacon commented:
Changelog:
 - Fixed object verbs not appearing
 - Added options to change between scrolling tabs and tabs wrapping to the new line
 - Sends only the tab data rather than literally all the verbs 4noraisin when we don't need to send to improve data lag.

Final Todo:
 - Fix OOC and Object mob verbs not showing up
 - Test for chat crashes

Commit:
epic games store

Commit:
Merge branch 'master' into TGUI-stat-panel

Commit:
Compilation

Commit:
Fixes verb overriding issue

Commit:
Significantly improves performance

Commit:
Update stat.dm

Commit:
Update mob_stat.dm

Commit:
Update mob_stat.dm

Commit:
stupid code

PowerfulBacon commented:
GetFlatIcon is just way too laggy to be useful

Commit:
fixes list being dumb

Commit:
Hashes the hash

Commit:
ike requests

ike709 commented on code/controllers/subsystem/stat.dm:
```diff
@@ -0,0 +1,70 @@
+			overlay_hash = "[overlay_hash][I.icon_state]"
+		what_to_search = "[M.type][M.name][overlay_hash]"
+	//Makes it shorter
+	what_to_search = sha1(what_to_search)
```
```suggestion
	what_to_search = md5(what_to_search)
```

My bad.

Commit:
Significantly improved image generation

PowerfulBacon commented:
GetFlatIcon has been replaced. There is now no spike in the subsystem when grabbing icons. 🎉

Commit:
Adds settings and stuff

Commit:
Lag improvements

Commit:
Update stat.dm

Commit:
Update stat.dm

park66665 commented on code/controllers/subsystem/stat.dm:
```diff
@@ -0,0 +1,75 @@
+//Whoever examins an item with a decal first, everyone else will see that items decals.
+//Significantly reduces server lag though, like MASSIVELY!
+/datum/controller/subsystem/stat/proc/get_flat_icon(atom/A)
+	var/what_to_search = "[A.type][A.icon_state[1]]"
```
```suggestion
	var/what_to_search = "[A.type][A.icon_state ? A.icon_state[1] : "*"]"
```

It turns out, tables, full tile windows and shits do have an empty string as `icon_state`. The whole proc runtimes and return when it tries to access the first letter of an empty string.

PowerfulBacon replied:
Yikes

park66665 commented on code/controllers/subsystem/stat.dm:
```diff
@@ -0,0 +1,75 @@
+		var/mob/M = A
+		var/overlay_hash = ""
+		for(var/image/I as() in M.overlays)
+			if(istext(I.icon_state))
```
```suggestion
			if(I.icon_state && istext(I.icon_state))
```

Gotta check here if `icon_state` is empty here too

PowerfulBacon replied:
Oh I thought istext catches null

PowerfulBacon committed:
Update code/controllers/subsystem/stat.dm

Co-authored-by: Xenomedes <park246824@gmail.com>

PowerfulBacon commented on tgui/packages/tgui-panel/stat/StatText.js:
```diff
@@ -0,0 +1,185 @@
+            params: params,
+          },
+        })}
+        color="transparent"
```
this should take into account the button colour option in settings.

Commit:
Merge branch 'TGUI-stat-panel' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork into TGUI-stat-panel

Commit:
Finishing touches

ike709 commented:
YOLO

francinum commented:
Why the fuck did you merge it in this state, what the fuck are you on?

VictorVow commented:
[RIP Linux](https://github.com/BeeStation/BeeStation-Hornet/issues/3363)
